File: modules/21b-evaluation-reliability/ts/uncertainty.ts
/**
 * uncertainty.ts — paired comparison, bootstrap CI, release verdict (Task 3).
 *
 * A byte-for-byte port of uncertainty.py: the same seeded LCG (full-product
 * modulo, high-bit multiply-shift index) and the same nearest-index percentile,
 * so the confidence interval is identical across languages.
 */

const TIE_EPS = 1e-9;
const LCG_A = 1664525;
const LCG_C = 1013904223;
const LCG_M = 4294967296; // 2 ** 32

/** Seeded LCG. State advances explicitly; no global RNG is touched. */
export class Lcg {
  state: number;

  constructor(seed: number) {
    this.state = seed >>> 0;
  }

  nextU32(): number {
    // (a*state + c) mod 2**32. The product stays below 2**53, so plain Number
    // arithmetic is exact — do NOT use Math.imul (32-bit wrap gives a different
    // value than Python's arbitrary-precision multiply-then-mod).
    this.state = (LCG_A * this.state + LCG_C) % LCG_M;
    return this.state;
  }

  randint(n: number): number {
    // High bits via multiply-shift; floor-divide (not >> 32, which is 32-bit).
    return Math.floor((this.nextU32() * n) / LCG_M);
  }
}

function sortedIndex(sortedValues: readonly number[], fractionalRank: number): number {
  let idx = Math.trunc(fractionalRank);
  if (idx < 0) idx = 0;
  if (idx > sortedValues.length - 1) idx = sortedValues.length - 1;
  return sortedValues[idx];
}

export function pairedBootstrapCi(
  diffs: readonly number[],
  iterations: number,
  seed: number,
  alpha = 0.05,
): [number, number] {
  const n = diffs.length;
  if (n === 0) throw new Error("diffs must be non-empty");
  if (iterations <= 0) throw new Error("iterations must be positive");
  const rng = new Lcg(seed);
  const means: number[] = [];
  for (let b = 0; b < iterations; b += 1) {
    let total = 0.0;
    for (let i = 0; i < n; i += 1) total += diffs[rng.randint(n)];
    means.push(total / n);
  }
  means.sort((x, y) => x - y);
  const lower = sortedIndex(means, Math.floor((alpha / 2.0) * iterations));
  const upper = sortedIndex(means, Math.ceil((1.0 - alpha / 2.0) * iterations) - 1);
  return [lower, upper];
}

export interface WinTieLoss {
  wins: number;
  ties: number;
  losses: number;
}

export function winTieLoss(
  baseline: readonly number[],
  candidate: readonly number[],
): WinTieLoss {
  if (baseline.length !== candidate.length) {
    throw new Error("score sequences must be the same length");
  }
  let wins = 0;
  let ties = 0;
  let losses = 0;
  for (let i = 0; i < baseline.length; i += 1) {
    const diff = candidate[i] - baseline[i];
    if (diff > TIE_EPS) wins += 1;
    else if (diff < -TIE_EPS) losses += 1;
    else ties += 1;
  }
  return { wins, ties, losses };
}

export interface Comparison {
  num_cases: number;
  mean_difference: number;
  ci_lower: number;
  ci_upper: number;
  alpha: number;
  practical_threshold: number;
  bootstrap_iterations: number;
  bootstrap_seed: number;
  win_tie_loss: WinTieLoss;
  verdict: "promote" | "reject" | "inconclusive";
}

export function compareVariants(
  baseline: readonly number[],
  candidate: readonly number[],
  practicalThreshold: number,
  iterations: number,
  seed: number,
  alpha = 0.05,
): Comparison {
  if (baseline.length !== candidate.length) {
    throw new Error("score sequences must be the same length");
  }
  const diffs = baseline.map((b, i) => candidate[i] - b);
  const meanDiff = diffs.reduce((a, b) => a + b, 0) / diffs.length;
  const [lower, upper] = pairedBootstrapCi(diffs, iterations, seed, alpha);
  let verdict: "promote" | "reject" | "inconclusive";
  if (lower >= practicalThreshold) verdict = "promote";
  else if (upper <= 0.0) verdict = "reject";
  else verdict = "inconclusive";
  return {
    num_cases: diffs.length,
    mean_difference: meanDiff,
    ci_lower: lower,
    ci_upper: upper,
    alpha,
    practical_threshold: practicalThreshold,
    bootstrap_iterations: iterations,
    bootstrap_seed: seed,
    win_tie_loss: winTieLoss(baseline, candidate),
    verdict,
  };
}

File: modules/21b-evaluation-reliability/py/uncertainty.py
"""uncertainty.py — paired comparison, bootstrap CI, release verdict (Task 3).

Compare two answer variants over the SAME held-out cases. Report the mean paired
difference, a paired bootstrap confidence interval (from scratch, with a SEEDED
deterministic resampler so it is reproducible and identical across languages),
win/tie/loss counts, and a verdict that can be ``inconclusive`` when the interval
crosses the practical-improvement threshold — so a model is not promoted on a
handful of noisy cases.
"""

from __future__ import annotations

import math
from collections.abc import Sequence
from typing import Any

_TIE_EPS = 1e-9

# Numerical Recipes linear congruential generator (mod 2**32). A tiny, fully
# specified PRNG so the resample sequence is reproducible and byte-identical in
# the TypeScript port.
_LCG_A = 1664525
_LCG_C = 1013904223
_LCG_M = 0x100000000


class Lcg:
    """Seeded LCG. State is advanced explicitly; no global RNG is touched."""

    def __init__(self, seed: int) -> None:
        self.state = seed & 0xFFFFFFFF

    def next_u32(self) -> int:
        self.state = (_LCG_A * self.state + _LCG_C) % _LCG_M
        return self.state

    def randint(self, n: int) -> int:
        # Use the HIGH bits (multiply-shift), not `% n`: an LCG's low bits are
        # highly periodic, so `next_u32() % n` degenerates for small n. The
        # product stays below 2**53, so it is exact in both Python and JS.
        return (self.next_u32() * n) >> 32


def _sorted_index(sorted_values: Sequence[float], fractional_rank: float) -> float:
    idx = int(fractional_rank)
    if idx < 0:
        idx = 0
    if idx > len(sorted_values) - 1:
        idx = len(sorted_values) - 1
    return sorted_values[idx]


def paired_bootstrap_ci(
    diffs: Sequence[float],
    iterations: int,
    seed: int,
    alpha: float = 0.05,
) -> tuple[float, float]:
    """95% (by default) paired bootstrap CI of the mean difference.

    Resamples ``diffs`` with replacement ``iterations`` times using a seeded LCG,
    then takes the lower/upper interval bounds from the sorted bootstrap means.
    """
    n = len(diffs)
    if n == 0:
        raise ValueError("diffs must be non-empty")
    if iterations <= 0:
        raise ValueError("iterations must be positive")
    rng = Lcg(seed)
    means: list[float] = []
    for _ in range(iterations):
        total = 0.0
        for _ in range(n):
            total += diffs[rng.randint(n)]
        means.append(total / n)
    means.sort()
    lower = _sorted_index(means, math.floor(alpha / 2.0 * iterations))
    upper = _sorted_index(means, math.ceil((1.0 - alpha / 2.0) * iterations) - 1)
    return lower, upper


def win_tie_loss(baseline: Sequence[float], candidate: Sequence[float]) -> dict[str, int]:
    """Per-case win/tie/loss counts for the candidate vs the baseline."""
    if len(baseline) != len(candidate):
        raise ValueError("score sequences must be the same length")
    wins = ties = losses = 0
    for b, c in zip(baseline, candidate, strict=True):
        diff = c - b
        if diff > _TIE_EPS:
            wins += 1
        elif diff < -_TIE_EPS:
            losses += 1
        else:
            ties += 1
    return {"wins": wins, "ties": ties, "losses": losses}


def compare_variants(
    baseline: Sequence[float],
    candidate: Sequence[float],
    practical_threshold: float,
    iterations: int,
    seed: int,
    alpha: float = 0.05,
) -> dict[str, Any]:
    """Full paired comparison + release verdict.

    Verdict:
      - ``promote``      when the CI lower bound clears the practical threshold,
      - ``reject``       when the CI upper bound is at or below 0,
      - ``inconclusive`` otherwise (the interval crosses the threshold).
    """
    if len(baseline) != len(candidate):
        raise ValueError("score sequences must be the same length")
    diffs = [c - b for b, c in zip(baseline, candidate, strict=True)]
    mean_diff = sum(diffs) / len(diffs)
    lower, upper = paired_bootstrap_ci(diffs, iterations, seed, alpha)
    if lower >= practical_threshold:
        verdict = "promote"
    elif upper <= 0.0:
        verdict = "reject"
    else:
        verdict = "inconclusive"
    return {
        "num_cases": len(diffs),
        "mean_difference": mean_diff,
        "ci_lower": lower,
        "ci_upper": upper,
        "alpha": alpha,
        "practical_threshold": practical_threshold,
        "bootstrap_iterations": iterations,
        "bootstrap_seed": seed,
        "win_tie_loss": win_tie_loss(baseline, candidate),
        "verdict": verdict,
    }
